README.md
SEUDoraemon
===========

Master Graduation Design


ChangeLog
2012-05-30 22:54 Zhao Bolu<zhaobl880704@gmail.com>
 V0.7
DoraemonV0.7
完成避障算法
2012-05-29 23:58 Zhao Bolu<zhaobl880704@gmail.com>
 V0.6
DoraemonV0.6
完成绕桩算法
其间因为两个类的互相引用一直报错，后来发现是由于编译器在预处理时，无法知道另一个类的大小导致的

2012-05-29 4:18 Zhao Bolu<zhaobl880704@gmail.com>
 V0.5
DoraemonV0.5
增加了底层控制，清理了不需要的代码

2012-05-28 14:30 Zhao Bolu<zhaobl880704@gmail.com>
 V0.4
DoraemonV0.4
使用openCV进行图像处理

2012-05-27 18:40 Zhao Bolu<zhaobl880704@gmail.com>
 V0.3
DoraemonV0.3
增加了camera pathplanning action ctrl模块

2012-05-26 18:40 Zhao Bolu<zhaobl880704@gmail.com>
 V0.2
DoraemonV0.2
调整代码结构
2012-05-24 18:40 Zhao Bolu<zhaobl880704@gmail.com>
 V0.1
DoraemonV0.1
研究生毕业论文

2013-05-31 18:40 Liu Xiangxiao<liuxx416@gmail.com>
V0.6.3
intergratedBioloidV0.6.3
毕业设计

2012-07-31 18:40 Zhao Bolu<zhaobl880704@gmail.com>
V0.6.2
intergratedBioloidV0.6.2
修正了main.cpp里陀螺仪传参问题。
修改了make install，可以将可执行码和data文件夹install到bin目录下

2012-07-31 15:50 Zhao Bolu<zhaobl880704@gmail.com>
V0.6.1
intergratedBioloidV0.6.1
矫正了陀螺仪

2012-07-31 12:30 Zhao Bolu<zhaobl880704@gmail.com>
V0.6
intergratedBioloidV0.6
增加了camera处理和转向处理

2012-07-31 11:00 Zhao Bolu<zhaobl880704@gmail.com>
V0.5.4
intergratedBioloidV0.5.4
修改了状态判断的函数

2012-07-31 10:00 Zhao Bolu<zhaobl880704@gmail.com> -
V0.5.3
intergratedBioloidV0.5.3
将bioloid模块替换为V0.7版本

2012-07-30 18:00 Zhao Bolu<zhaobl880704@gmail.com>
V0.5.2
intergratedBioloidV0.5.2
修改了main.cpp，将初始化放到主线程里，因为发现串口和陀螺仪同时初始化会报段错误

2012-07-30 18:00 Zhao Bolu<zhaobl880704@gmail.com>
V0.5.1
intergratedBioloidV0.5.1
添加了data文件夹，程序执行时，可执行码与data文件夹应放在同一目录下

2012-07-30 17:40 Zhao Bolu<zhaobl880704@gmail.com>
V0.5
intergratedBioloidV0.5
增加了陀螺仪模块rmg

2012-07-30 15:30 Zhao Bolu<zhaobl880704@gmail.com>
V0.4.2
intergratedBioloidV0.4.2
修改了AX12_C3.cpp里RegSetPosition和Action函数中信息阻塞的代码

2012-07-30 14:00 Zhao Bolu<zhaobl880704@gmail.com>
V0.4.1
intergratedBioloidV0.4.1
修改了main.cpp里忘记初始化com3的bug

2012-07-24 20:00 Zhao Bolu<zhaobl880704@gmail.com>
V0.4
intergratedBioloidV0.4
将com模块更新，使之不会被未接收到舵机反馈信息而阻塞
2012-07-21 23:20 Zhao Bolu<zhaobl880704@gmail.com>
V0.3
intergratedBioloidV0.3
增加了camera模块，由于该模块是用C写的，C++调用时，注意添加
#ifdef __cplusplus
extern "C" {
#endif
/**
*C的定义
*
*/
#ifdef __cplusplus
}
#endif
这样的定义
2012-07-21 23:20 Zhao Bolu<zhaobl880704@gmail.com>
V0.2
intergratedBioloidV0.2
调整了RoBoIO的文件位置结构

2012-07-21 23:20 Zhao Bolu<zhaobl880704@gmail.com>
V0.1
intergratedBioloidV0.1
将串口通讯模块com加入代码

2012-07-21 18:20 Zhao Bolu<zhaobl880704@gmail.com>
V0.5
corebioloidV0.5
将主程序改成多线程，使用pthread线程库
现在主要为三个线程，姿态传感器线程，摄像头线程和控制线程
增加设置控制周期的接口，但建议只调用一次

2012-07-21 9:20 Zhao Bolu<zhaobl880704@gmail.com>
V0.4
corebioloidV0.4
增加了根据身体状态调整使用序列的方法，这个应该是比赛的主要使用模式
2012-07-19 14:00 Zhao Bolu<zhaobl880704@gmail.com>
corebioloidV0.3
修改了一些类变量，加了一些注释
2012-07-10 16:00 Zhao Bolu<zhaobl880704@gmail.com>
corebioloidV0.2
修改了一些部分,考虑减小数据内存使用
2012-07-10 13:30 Zhao Bolu<zhaobl880704@gmail.com>
corebioloidV0.1
bioloid核心模块控制程序
main.cpp是测试程序
bioloid目录下是模块代码
rsg下为动作序列的机器人模型


src/com/IMU.cpp
#ifndef __IMU_H
#define __IMU_H
/* 
编写者：lisn3188
网址：www.chiplab7.com
作者E-mail：lisn3188@163.com
编译环境：MDK-Lite  Version: 4.23
初版时间: 2012-05-25
测试： 本程序已在第七实验室的mini IMU上完成测试
功能：
使用串行中断程序接收mini IMU 的数据。
移植者需在自己编写对应的串口初始化程序，并开启接收中断。
在中断子程序中 需要做相应的改动

为了使用这个程序，你需要：
1.初始化你的串口，将波特率与模块的波特率设置成一致的
2.开启串口的接收中断，并参考以下程序编写中断子程序
  参考 void USART2_IRQHandler(void) 
3.在主程序的循环体中，定时调用子程序：
  void UART2_CommandRoute(void)
  以检查是否收到新的数据帧
  所有的姿态数据和ADC值都会在 UART2_CommandRoute
  中进行更新。
4.使用 以下值会自动更新：

float 	imu_calc_yaw,  //偏航角
		imu_calc_pitch,//俯仰
		imu_calc_roll, //滚转
		imu_calc_alt,  //高度
		imu_calc_tempr,//温度
		imu_calc_press;//气压
int16_t imu_orig_ax, imu_orig_ay, imu_orig_az;//加速度计
int16_t imu_orig_gx, imu_orig_gy, imu_orig_gz;//陀螺仪
int16_t imu_orig_hx, imu_orig_hy, imu_orig_hz;//磁力计
------------------------------------
*/
#include <stdio.h>
#include "roboard.h"
//uart reicer flag
#define b_uart_head  0x80  //收到A5 头 标志位
#define b_rx_over    0x40  //收到完整的帧标志
// USART Receiver buffer
#define RX_BUFFER_SIZE 100 //接收缓冲区字节数

volatile unsigned char rx_buffer[RX_BUFFER_SIZE]; //接收数据缓冲区
volatile unsigned char rx_wr_index; //缓冲写指针
volatile unsigned char RC_Flag;  //接收状态标志字节
//typedef int int16_t;
float 	imu_calc_yaw,  //偏航角
		imu_calc_pitch,//俯仰
		imu_calc_roll, //滚转
		imu_calc_alt,  //高度
		imu_calc_tempr,//温度
		imu_calc_press;//气压
//在接收完一帧IMU姿态报告后，调用这个子程序来取出姿态数据
void UART2_Get_IMU(void)
{
	int16_t temp;
	
	temp = 0;
	temp = rx_buffer[2];
	temp <<= 8;
	temp |= rx_buffer[3];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_calc_yaw=(float)temp / 10.0f; //偏航角
	
	temp = 0;
	temp = rx_buffer[4];
	temp <<= 8;
	temp |= rx_buffer[5];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_calc_pitch=(float)temp / 10.0f;//俯仰
	
	temp = 0;
	temp = rx_buffer[6];
	temp <<= 8;
	temp |= rx_buffer[7];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_calc_roll=(float)temp / 10.0f;//滚转

	temp = 0;
	temp = rx_buffer[8];
	temp <<= 8;
	temp |= rx_buffer[9];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_calc_alt=(float)temp / 10.0f;//高度
	
	temp = 0;
	temp = rx_buffer[10];
	temp <<= 8;
	temp |= rx_buffer[11];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_calc_tempr=(float)temp / 10.0f;//温度
	
	temp = 0;
	temp = rx_buffer[12];
	temp <<= 8;
	temp |= rx_buffer[13];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_calc_press=(float)temp * 10.0f;//气压

}

//ADC值
int16_t imu_orig_ax, imu_orig_ay, imu_orig_az;//加速度计
int16_t imu_orig_gx, imu_orig_gy, imu_orig_gz;//陀螺仪
int16_t imu_orig_hx, imu_orig_hy, imu_orig_hz;//磁力计
//在接收一帧ReportMotion 后调用这个子程序来取出ADC数据
void UART2_Get_Motion(void)
{
	int16_t temp;
	
	temp = 0;
	temp = rx_buffer[2];
	temp <<= 8;
	temp |= rx_buffer[3];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_orig_ax=temp;//加速度计 X轴的ADC值
	
	temp = 0;
	temp = rx_buffer[4];
	temp <<= 8;
	temp |= rx_buffer[5];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_orig_ay=temp;//加速度计 Y轴的ADC值
	
	temp = 0;
	temp = rx_buffer[6];
	temp <<= 8;
	temp |= rx_buffer[7];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_orig_az=temp;//加速度计 Z轴的ADC值
	
	temp = 0;
	temp = rx_buffer[8];
	temp <<= 8;
	temp |= rx_buffer[9];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_orig_gx=temp;//陀螺仪 X轴的ADC值
	
	temp = 0;
	temp = rx_buffer[10];
	temp <<= 8;
	temp |= rx_buffer[11];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_orig_gy=temp;//陀螺仪 Y轴的ADC值
	
	temp = 0;
	temp = rx_buffer[12];
	temp <<= 8;
	temp |= rx_buffer[13];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_orig_gz=temp;//陀螺仪 Z轴的ADC值
	
	temp = 0;
	temp = rx_buffer[14];
	temp <<= 8;
	temp |= rx_buffer[15];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_orig_hx=temp;//磁力计 X轴的ADC值
	
	temp = 0;
	temp = rx_buffer[16];
	temp <<= 8;
	temp |= rx_buffer[17];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_orig_hy=temp;//磁力计 Y轴的ADC值
	
	temp = 0;
	temp = rx_buffer[18];
	temp <<= 8;
	temp |= rx_buffer[19];
	if(temp&0x8000){
	temp = 0-(temp&0x7fff);
	}else temp = (temp&0x7fff);
	imu_orig_hz=temp;//磁力计 Z轴的ADC值
}


//以下中断子程序是基于STM32的，如果你使用的不是STM32 请做相应的修改
//------------------------------------------------------
void USART2_IRQHandler(void) //接收中断子程序
{
  unsigned char data;
  data=(unsigned char)com4_Read(); //读取接收到的字节
  if(data==0xa5) 
  { 
      RC_Flag|=b_uart_head; //如果接收到A5 置位帧头标专位
      rx_buffer[rx_wr_index++]=data; //保存这个字节.
  }
  else if(data==0x5a)
       { 
	   if(RC_Flag&b_uart_head) //如果上一个字节是A5 那么认定 这个是帧起始字节
	     { rx_wr_index=0;  //重置 缓冲区指针
		   RC_Flag&=~b_rx_over; //这个帧才刚刚开始收
         }
         else //上一个字节不是A5
		  rx_buffer[rx_wr_index++]=data;
         RC_Flag&=~b_uart_head; //清帧头标志
       }
	   else
	   { rx_buffer[rx_wr_index++]=data;
		 RC_Flag&=~b_uart_head;
		 if(rx_wr_index==rx_buffer[0]) //收够了字节数.
	     {  
			RC_Flag|=b_rx_over; //置位 接收完整的一帧数据
          }
	   }

  if(rx_wr_index==RX_BUFFER_SIZE) //防止缓冲区溢出
  rx_wr_index--;
}

//--校验当前接收到的一帧数据是否 与帧校验字节一致
unsigned char Sum_check(void)
{ 
  unsigned char i;
  int16_t checksum=0; 
  for(i=0;i<rx_buffer[0]-2;i++)
   checksum+=rx_buffer[i];
  if((checksum%256)==rx_buffer[rx_buffer[0]-2])
   return(0x01); //Checksum successful
  else
   return(0x00); //Checksum error
}
bool imu_is_update_imu;
bool imu_is_update_motion;

//--这个子程序需要在主程序中 定时调用,以检查 串口是否接收完一帧数据
void UART2_CommandRoute(void)
{
     imu_is_update_imu=false;
 imu_is_update_motion=false;
  if(RC_Flag&b_rx_over){  //已经接收完一帧?
	RC_Flag&=~b_rx_over; //清标志先
	if(Sum_check()){ 
	//校验通过
	  if(rx_buffer[1]==0xA1){ //UART2_ReportIMU 的数据
// 	    printf("get IMU succ\n");
              imu_is_update_imu=true;
	    UART2_Get_IMU();	//取数据
	  }
	  if(rx_buffer[1]==0xA2){ //UART2_ReportMotion 的数据
// 	    printf("get Motion succ\n");
              imu_is_update_motion=true;
	    UART2_Get_Motion();	 //取数据
	  } 
	
	}//校验是否通过?
  }
}

#endif

